chpl/chpl-api/src/main/java/gov/healthit/chpl/registration/RateLimitingInterceptor.java

```diff
@@ -0,0 +1,127 @@
+package gov.healthit.chpl.registration;
+
+import gov.healthit.chpl.dao.ApiKeyDAO;
+import gov.healthit.chpl.dto.ApiKeyDTO;
+import gov.healthit.chpl.util.Util;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.EnvironmentAware;
+import org.springframework.context.MessageSource;
+import org.springframework.context.i18n.LocaleContextHolder;
+import org.springframework.context.support.DefaultMessageSourceResolvable;
+import org.springframework.core.env.Environment;
+import org.springframework.http.HttpStatus;
+import org.springframework.stereotype.Component;
+import org.springframework.util.StringUtils;
+import org.springframework.web.servlet.handler.HandlerInterceptorAdapter;
+
+@Component
+public class RateLimitingInterceptor extends HandlerInterceptorAdapter implements EnvironmentAware {
+
+    private static final Logger LOGGER = LogManager.getLogger(RateLimitingInterceptor.class);
+
+    @Autowired
+    private Environment env;
+
+    @Autowired
+    private MessageSource messageSource;
+
+    @Autowired
+    private ApiKeyDAO apiKeyDao;
+
+    private String timeUnit;
+
+    private int limit;
+
+    private Map<String, SimpleRateLimiter> limiters = new ConcurrentHashMap<>();
+
+    private List<String> whitelist = new ArrayList<String>();
+
+    public RateLimitingInterceptor() {
+    }
+
+    @Override
+    public void setEnvironment(Environment env) {
+        LOGGER.info("setEnvironment");
+        this.env = env;
+        this.timeUnit = env.getProperty("rateLimitTimeUnit");
+        this.limit = Integer.valueOf(env.getProperty("rateTokenLimit"));
+    }
+
+    /** {@inheritDoc} */
+    public String getMessage(final String messageCode, final String input, final String input2) {
```

Comment:
You have Katy's newest code, right? And in that I think she had a utility function for this functionality. Should use that

Reply:
The class is ErrorMessageUtil, you can see examples of it being used in the gov.healthit.chpl.validation.listing.reviewer code